File: apps/ios/Ash/Ash/Core/Services/PadManager.swift
//
//  PadManager.swift
//  Ash
//
//  Core Service - Manages pad state using Rust core for allocation logic
//
//  This service wraps the Rust Pad implementation to ensure iOS and Android
//  use identical pad allocation logic. All pad operations go through this service.
//

import Foundation

// MARK: - Sendable Conformance for FFI Types

// Role enum from UniFFI is safe to send across concurrency boundaries
extension Role: @unchecked Sendable {}

/// Errors that can occur during pad operations
enum PadManagerError: Error, Sendable {
    case padNotFound
    case insufficientBytes
    case storageError(Error)
    case invalidState
}

/// Protocol for pad management operations
/// Note: loadPad is not in protocol as Pad is non-Sendable (implementation detail)
protocol PadManagerProtocol: Sendable {
    /// Store pad bytes with initial state (after ceremony)
    func storePad(bytes: [UInt8], for conversationId: String) async throws

    /// Check if a message of given length can be sent
    func canSend(length: UInt32, role: Role, for conversationId: String) async throws -> Bool

    /// Get bytes available for sending
    func availableForSending(role: Role, for conversationId: String) async throws -> UInt64

    /// Consume pad bytes for sending a message
    /// Returns the key bytes for encryption
    func consumeForSending(length: UInt32, role: Role, for conversationId: String) async throws -> [UInt8]

    /// Get the next send offset (for message sequencing)
    func nextSendOffset(role: Role, for conversationId: String) async throws -> UInt64

    /// Update peer's consumption based on received message
    func updatePeerConsumption(peerRole: Role, consumed: UInt64, for conversationId: String) async throws

    /// Get pad bytes for decryption at a specific offset
    func getBytesForDecryption(offset: UInt64, length: UInt64, for conversationId: String) async throws -> [UInt8]

    /// Get current pad state (for UI display)
    func getPadState(for conversationId: String) async throws -> PadState

    /// Zero pad bytes at specific offset (for forward secrecy)
    /// When a message expires, the key material is zeroed to prevent future decryption
    func zeroPadBytes(offset: UInt64, length: UInt64, for conversationId: String) async throws

    /// Wipe pad for a conversation
    func wipePad(for conversationId: String) async throws

    /// Wipe all pads
    func wipeAllPads() async throws

    /// Get pad bytes for token derivation (auth/burn tokens)
    /// Must be called BEFORE wiping the pad
    func getPadBytes(for conversationId: String) async throws -> [UInt8]
}

/// Represents the current state of a pad
struct PadState: Sendable {
    let totalBytes: UInt64
    let consumedFront: UInt64
    let consumedBack: UInt64
    let remaining: UInt64
    let isExhausted: Bool
}

/// Data stored in Keychain for a pad
struct PadStorageData: Codable {
    let bytes: [UInt8]
    let consumedFront: UInt64
    let consumedBack: UInt64

    func encode() throws -> Data {
        try JSONEncoder().encode(self)
    }

    static func decode(from data: Data) throws -> PadStorageData {
        try JSONDecoder().decode(PadStorageData.self, from: data)
    }
}

/// Actor managing pad state using Rust core
actor PadManager: PadManagerProtocol {
    private let keychainService: KeychainServiceProtocol

    /// In-memory cache of loaded pads
    private var padCache: [String: Pad] = [:]

    init(keychainService: KeychainServiceProtocol = KeychainService()) {
        self.keychainService = keychainService
    }

    // MARK: - Load/Store

    func loadPad(for conversationId: String) async throws -> Pad {
        // Check cache first
        if let cached = padCache[conversationId] {
            return cached
        }

        // Load from Keychain
        guard let data = try keychainService.retrieve(for: keychainKey(for: conversationId)) else {
            throw PadManagerError.padNotFound
        }

        let storageData = try PadStorageData.decode(from: data)

        // Create Rust Pad with state
        let pad = Pad.fromBytesWithState(
            bytes: storageData.bytes,
            consumedFront: storageData.consumedFront,
            consumedBack: storageData.consumedBack
        )

        padCache[conversationId] = pad
        return pad
    }

    func storePad(bytes: [UInt8], for conversationId: String) async throws {
        let storageData = PadStorageData(
            bytes: bytes,
            consumedFront: 0,
            consumedBack: 0
        )

        let encoded = try storageData.encode()
        try keychainService.store(data: encoded, for: keychainKey(for: conversationId))

        // Create and cache the Rust Pad
        let pad = Pad.fromBytes(bytes: bytes)
        padCache[conversationId] = pad
    }

    private func savePadState(pad: Pad, for conversationId: String) throws {
        let storageData = PadStorageData(
            bytes: pad.asBytes(),
            consumedFront: pad.consumedFront(),
            consumedBack: pad.consumedBack()
        )

        let encoded = try storageData.encode()
        try keychainService.store(data: encoded, for: keychainKey(for: conversationId))
    }

    // MARK: - Send Operations

    func canSend(length: UInt32, role: Role, for conversationId: String) async throws -> Bool {
        let pad = try await loadPad(for: conversationId)
        return pad.canSend(length: length, role: role)
    }

    func availableForSending(role: Role, for conversationId: String) async throws -> UInt64 {
        let pad = try await loadPad(for: conversationId)
        return pad.availableForSending(role: role)
    }

    func consumeForSending(length: UInt32, role: Role, for conversationId: String) async throws -> [UInt8] {
        let pad = try await loadPad(for: conversationId)

        // Consume bytes using Rust Pad
        let keyBytes = try pad.consume(n: length, role: role)

        // Persist updated state
        try savePadState(pad: pad, for: conversationId)

        return keyBytes
    }

    func nextSendOffset(role: Role, for conversationId: String) async throws -> UInt64 {
        let pad = try await loadPad(for: conversationId)
        return pad.nextSendOffset(role: role)
    }

    // MARK: - Receive Operations

    func updatePeerConsumption(peerRole: Role, consumed: UInt64, for conversationId: String) async throws {
        let pad = try await loadPad(for: conversationId)
        pad.updatePeerConsumption(peerRole: peerRole, newConsumed: consumed)

        // Persist updated state
        try savePadState(pad: pad, for: conversationId)
    }

    func getBytesForDecryption(offset: UInt64, length: UInt64, for conversationId: String) async throws -> [UInt8] {
        let pad = try await loadPad(for: conversationId)
        let bytes = pad.asBytes()

        let start = Int(offset)
        let end = min(start + Int(length), bytes.count)

        guard start >= 0, end <= bytes.count, start < end else {
            throw PadManagerError.insufficientBytes
        }

        return Array(bytes[start..<end])
    }

    // MARK: - State Queries

    func getPadState(for conversationId: String) async throws -> PadState {
        let pad = try await loadPad(for: conversationId)
        return PadState(
            totalBytes: pad.totalSize(),
            consumedFront: pad.consumedFront(),
            consumedBack: pad.consumedBack(),
            remaining: pad.remaining(),
            isExhausted: pad.isExhausted()
        )
    }

    // MARK: - Forward Secrecy

    func zeroPadBytes(offset: UInt64, length: UInt64, for conversationId: String) async throws {
        let pad = try await loadPad(for: conversationId)

        // Zero the bytes using Rust's secure zeroing
        let success = pad.zeroBytesAt(offset: offset, length: length)

        if success {
            // Persist updated state (with zeroed bytes)
            try savePadState(pad: pad, for: conversationId)
            Log.debug(.crypto, "Zeroed \(length) pad bytes at offset \(offset) for forward secrecy")
        } else {
            Log.warning(.crypto, "Failed to zero pad bytes: offset \(offset), length \(length) out of bounds")
        }
    }

    // MARK: - Cleanup

    func wipePad(for conversationId: String) async throws {
        try keychainService.delete(for: keychainKey(for: conversationId))
        padCache.removeValue(forKey: conversationId)
    }

    func wipeAllPads() async throws {
        // Delete all pad keys from keychain
        // Note: This assumes we can identify pad keys by prefix
        try keychainService.deleteAll()
        padCache.removeAll()
    }

    func getPadBytes(for conversationId: String) async throws -> [UInt8] {
        let pad = try await loadPad(for: conversationId)
        return pad.asBytes()
    }

    // MARK: - Helpers

    private func keychainKey(for conversationId: String) -> String {
        "pad.\(conversationId)"
    }
}

File: apps/android/app/src/main/java/com/monadial/ash/core/services/PadManager.kt
package com.monadial.ash.core.services

import android.util.Log
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import uniffi.ash.Pad
import uniffi.ash.Role
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Pad state for UI display
 */
data class PadState(
    val totalBytes: Long,
    val consumedFront: Long,
    val consumedBack: Long,
    val remaining: Long,
    val isExhausted: Boolean
)

/**
 * PadManager - Manages pad state using Rust core for allocation logic
 *
 * This service wraps the Rust Pad implementation to ensure iOS and Android
 * use identical pad allocation logic. All pad operations go through this service.
 *
 * Matching iOS: apps/ios/Ash/Ash/Core/Services/PadManager.swift
 */
@Singleton
class PadManager @Inject constructor(
    private val conversationStorage: ConversationStorageService
) {
    companion object {
        private const val TAG = "PadManager"
    }

    // Mutex for thread-safe access (replaces @Synchronized for suspend functions)
    private val mutex = Mutex()

    // In-memory cache of loaded pads (Rust Pad objects)
    private val padCache = mutableMapOf<String, Pad>()

    // MARK: - Load/Store

    /**
     * Load pad for a conversation, using cache if available.
     * Matching iOS: PadManager.loadPad
     */
    suspend fun loadPad(conversationId: String): Pad = mutex.withLock {
        // Check cache first
        padCache[conversationId]?.let { return@withLock it }

        // Load from storage (matching iOS: PadStorageData from Keychain)
        val storageData = conversationStorage.getPadStorageData(conversationId)
            ?: throw IllegalStateException("Pad not found for conversation $conversationId")

        val padBytes = android.util.Base64.decode(storageData.bytes, android.util.Base64.NO_WRAP)

        // Create Rust Pad with state (matching iOS: Pad.fromBytesWithState)
        val pad = Pad.fromBytesWithState(
            padBytes.map { it.toUByte() },
            storageData.consumedFront.toULong(),
            storageData.consumedBack.toULong()
        )

        Log.d(TAG, "Loaded pad for ${conversationId.take(8)}: front=${storageData.consumedFront}, back=${storageData.consumedBack}")

        padCache[conversationId] = pad
        pad
    }

    /**
     * Store pad bytes for a new conversation (after ceremony)
     */
    suspend fun storePad(bytes: ByteArray, conversationId: String) = mutex.withLock {
        conversationStorage.savePadBytes(conversationId, bytes)

        // Create and cache the Rust Pad
        val pad = Pad.fromBytes(bytes.map { it.toUByte() })
        padCache[conversationId] = pad
    }

    /**
     * Save current pad state to storage.
     * Matching iOS: PadManager.savePadState - saves bytes + consumption together
     * Note: Called within mutex.withLock, no additional locking needed
     */
    private suspend fun savePadState(pad: Pad, conversationId: String) {
        val consumedFront = pad.consumedFront().toLong()
        val consumedBack = pad.consumedBack().toLong()
        val padBytes = pad.asBytes().map { it.toByte() }.toByteArray()

        // Save bytes + consumption state together (matching iOS PadStorageData)
        conversationStorage.savePadState(
            conversationId = conversationId,
            padBytes = padBytes,
            consumedFront = consumedFront,
            consumedBack = consumedBack
        )
    }

    // MARK: - Send Operations

    /**
     * Check if a message of given length can be sent
     */
    suspend fun canSend(length: Int, role: Role, conversationId: String): Boolean {
        val pad = loadPad(conversationId)
        return pad.canSend(length.toUInt(), role)
    }

    /**
     * Get bytes available for sending
     */
    suspend fun availableForSending(role: Role, conversationId: String): Long {
        val pad = loadPad(conversationId)
        return pad.availableForSending(role).toLong()
    }

    /**
     * Get the next send offset (for message sequencing)
     */
    suspend fun nextSendOffset(role: Role, conversationId: String): Long {
        val pad = loadPad(conversationId)
        return pad.nextSendOffset(role).toLong()
    }

    /**
     * Consume pad bytes for sending a message.
     * Returns the key bytes for encryption.
     *
     * IMPORTANT: This updates consumption state - call only once per message!
     */
    suspend fun consumeForSending(length: Int, role: Role, conversationId: String): ByteArray = mutex.withLock {
        // Get cached pad or load it (matching iOS pattern)
        val pad = padCache[conversationId] ?: run {
            val storageData = conversationStorage.getPadStorageData(conversationId)
                ?: throw IllegalStateException("Pad not found for conversation $conversationId")
            val padBytes = android.util.Base64.decode(storageData.bytes, android.util.Base64.NO_WRAP)
            Pad.fromBytesWithState(
                padBytes.map { it.toUByte() },
                storageData.consumedFront.toULong(),
                storageData.consumedBack.toULong()
            ).also { padCache[conversationId] = it }
        }

        Log.d(TAG, "Consuming $length bytes for sending (role=$role, conv=${conversationId.take(8)})")

        // Consume bytes using Rust Pad
        val keyBytes = pad.consume(length.toUInt(), role)

        // Persist updated state
        savePadState(pad, conversationId)

        Log.d(TAG, "Consumption complete. Front=${pad.consumedFront()}, Back=${pad.consumedBack()}")

        keyBytes.map { it.toByte() }.toByteArray()
    }

    // MARK: - Receive Operations

    /**
     * Update peer's consumption based on received message
     */
    suspend fun updatePeerConsumption(peerRole: Role, consumed: Long, conversationId: String) = mutex.withLock {
        // Get cached pad or load it (matching iOS pattern)
        val pad = padCache[conversationId] ?: run {
            val storageData = conversationStorage.getPadStorageData(conversationId)
                ?: throw IllegalStateException("Pad not found for conversation $conversationId")
            val padBytes = android.util.Base64.decode(storageData.bytes, android.util.Base64.NO_WRAP)
            Pad.fromBytesWithState(
                padBytes.map { it.toUByte() },
                storageData.consumedFront.toULong(),
                storageData.consumedBack.toULong()
            ).also { padCache[conversationId] = it }
        }

        pad.updatePeerConsumption(peerRole, consumed.toULong())

        // Persist updated state
        savePadState(pad, conversationId)

        Log.d(TAG, "Updated peer consumption. Front=${pad.consumedFront()}, Back=${pad.consumedBack()}")
    }

    /**
     * Get pad bytes for decryption at a specific offset
     */
    suspend fun getBytesForDecryption(offset: Long, length: Int, conversationId: String): ByteArray {
        val pad = loadPad(conversationId)
        val bytes = pad.asBytes()

        val start = offset.toInt()
        val end = minOf(start + length, bytes.size)

        if (start < 0 || end > bytes.size || start >= end) {
            throw IllegalStateException("Invalid pad range: offset=$offset, length=$length, padSize=${bytes.size}")
        }

        return bytes.subList(start, end).map { it.toByte() }.toByteArray()
    }

    // MARK: - State Queries

    /**
     * Get current pad state (for UI display)
     */
    suspend fun getPadState(conversationId: String): PadState {
        val pad = loadPad(conversationId)
        return PadState(
            totalBytes = pad.totalSize().toLong(),
            consumedFront = pad.consumedFront().toLong(),
            consumedBack = pad.consumedBack().toLong(),
            remaining = pad.remaining().toLong(),
            isExhausted = pad.isExhausted()
        )
    }

    // MARK: - Forward Secrecy

    /**
     * Zero pad bytes at specific offset (for forward secrecy).
     * When a message expires, the key material is zeroed to prevent future decryption.
     */
    suspend fun zeroPadBytes(offset: Long, length: Int, conversationId: String) = mutex.withLock {
        // Get cached pad or load it (matching iOS pattern)
        val pad = padCache[conversationId] ?: run {
            val storageData = conversationStorage.getPadStorageData(conversationId)
                ?: throw IllegalStateException("Pad not found for conversation $conversationId")
            val padBytes = android.util.Base64.decode(storageData.bytes, android.util.Base64.NO_WRAP)
            Pad.fromBytesWithState(
                padBytes.map { it.toUByte() },
                storageData.consumedFront.toULong(),
                storageData.consumedBack.toULong()
            ).also { padCache[conversationId] = it }
        }

        val success = pad.zeroBytesAt(offset.toULong(), length.toULong())

        if (success) {
            // Persist updated state (with zeroed bytes) - matching iOS savePadState
            savePadState(pad, conversationId)
            Log.d(TAG, "Zeroed $length pad bytes at offset $offset for forward secrecy")
        } else {
            Log.w(TAG, "Failed to zero pad bytes: offset $offset, length $length out of bounds")
        }
    }

    // MARK: - Cleanup

    /**
     * Wipe pad for a conversation
     */
    suspend fun wipePad(conversationId: String) = mutex.withLock {
        padCache.remove(conversationId)
        conversationStorage.deletePadBytes(conversationId)
    }

    /**
     * Get pad bytes (for token derivation before wiping)
     */
    suspend fun getPadBytes(conversationId: String): ByteArray {
        val pad = loadPad(conversationId)
        return pad.asBytes().map { it.toByte() }.toByteArray()
    }

    /**
     * Invalidate cached pad (call when conversation is deleted)
     */
    @Synchronized
    fun invalidateCache(conversationId: String) {
        padCache.remove(conversationId)
    }

    /**
     * Clear all cached pads
     */
    @Synchronized
    fun clearCache() {
        padCache.clear()
    }
}
